samples/bluetooth/broadcast_audio_sink/src/main.c

```diff
@@ -83,99 +122,221 @@ static uint8_t sink_broadcast_code[BT_AUDIO_BROADCAST_CODE_SIZE];
 
-#if defined(CONFIG_LIBLC3)
+#if defined(CONFIG_USB_DEVICE_AUDIO)
+int16_t usb_audio_data[MAX_NUM_SAMPLES_MONO] = {0};
+int16_t usb_audio_data_stereo[MAX_NUM_SAMPLES_STEREO] = {0};
 
-#include "lc3.h"
+RING_BUF_DECLARE(ring_buf_usb, AUDIO_RING_BUF_SIZE);
+NET_BUF_POOL_DEFINE(usb_tx_buf_pool, USB_TX_BUF_NUM, BROADCAST_STEREO_SAMPLE_SIZE, 0,
+		    net_buf_destroy);
 
-#define MAX_SAMPLE_RATE       16000
-#define MAX_FRAME_DURATION_US 10000
-#define MAX_NUM_SAMPLES       ((MAX_FRAME_DURATION_US * MAX_SAMPLE_RATE) / USEC_PER_SEC)
+static void mix_mono_to_stereo(void);
+#endif /* defined(CONFIG_USB_DEVICE_AUDIO) */
 
-static int16_t audio_buf[MAX_NUM_SAMPLES];
-static lc3_decoder_t lc3_decoder;
-static lc3_decoder_mem_16k_t lc3_decoder_mem;
+#if defined(CONFIG_LIBLC3)
+static int16_t audio_buf[MAX_NUM_SAMPLES_MONO];
 static int frames_per_sdu;
+static K_SEM_DEFINE(lc3_decoder_sem, 0, 1);
+
+static void do_lc3_decode(struct broadcast_sink_stream *sink_stream);
+static void lc3_decoder_thread(void *arg1, void *arg2, void *arg3);
+K_THREAD_DEFINE(decoder_tid, LC3_ENCODER_STACK_SIZE, lc3_decoder_thread,
+		NULL, NULL, NULL, LC3_ENCODER_PRIORITY, 0, -1);
+
+/* Consumer thread of the decoded stream data */
+static void lc3_decoder_thread(void *arg1, void *arg2, void *arg3)
+{
+	while (true) {
+		k_sem_take(&lc3_decoder_sem, K_FOREVER);
+		/* For now we only handle one channel, so always only decode the first */
+		do_lc3_decode(&streams[0]);
+
+#if defined(CONFIG_USB_DEVICE_AUDIO)
+		/* If the ring buffer usage is alrger than zero, then there is data to process */
+		if (ring_buf_space_get(&streams[0].audio_ring_buf)) {
+			mix_mono_to_stereo();
+		}
+#endif /* #if defined(CONFIG_USB_DEVICE_AUDIO) */
+	}
+}
+
+static void do_lc3_decode(struct broadcast_sink_stream *sink_stream)
+{
+	int err = 0;
+	int offset = 0;
+	uint8_t *buf_data;
+	struct net_buf *ptr_net_buf;
+	int octets_per_frame;
+
+	k_mutex_lock(&sink_stream->lc3_decoder_mutex, K_FOREVER);
+
+	if (sink_stream->in_buf == NULL) {
+		k_mutex_unlock(&sink_stream->lc3_decoder_mutex);
+		return;
+	}
+
+	ptr_net_buf = net_buf_ref(sink_stream->in_buf);
+	net_buf_unref(sink_stream->in_buf);
+	sink_stream->in_buf = NULL;
+	k_mutex_unlock(&sink_stream->lc3_decoder_mutex);
 
-static int lc3_enable(const struct bt_audio_codec_cfg *codec_cfg)
+	buf_data = ptr_net_buf->data;
+	octets_per_frame = ptr_net_buf->len / frames_per_sdu;
+
+	for (int i = 0; i < frames_per_sdu; i++) {
+		err = lc3_decode(sink_stream->lc3_decoder, buf_data + offset, octets_per_frame,
+				 LC3_PCM_FORMAT_S16, audio_buf, 1);
+
+		if (err == 1) {
+			printk("  decoder performed PLC\n");
+		} else if (err < 0) {
+			printk("  decoder failed - wrong parameters? (err = %d)\n", err);
+		}
+
+		offset += octets_per_frame;
+	}
+
+	net_buf_unref(ptr_net_buf);
+
+#if defined(CONFIG_USB_DEVICE_AUDIO)
+	uint32_t rbret;
+
+	if (ring_buf_space_get(&sink_stream->audio_ring_buf) == 0) {
+		ring_buf_reset(&sink_stream->audio_ring_buf);
+	}
```

Comment:
Can you elaborate (in a code comment perhaps?) why we reset the buffer when there is no more space left?